pdg/src/builder.rs

```diff
@@ -1,136 +1,148 @@
-        EventKind::AddrOfLocal(lhs, _) => *lhs,
-        EventKind::Offset(ptr, _, _) => *ptr,
-        EventKind::Done => return None,
-    })
+pub trait EventKindExt {
+    fn ptr(&self, metadata: &EventMetadata) -> Option<Pointer>;
+    fn has_parent(&self) -> bool;
+    fn parent(&self, obj: (GraphId, NodeId)) -> Option<(GraphId, NodeId)>;
+    fn to_node_kind(&self) -> Option<NodeKind>;
 }
 
-fn get_parent_object(kind: &EventKind, obj: (GraphId, NodeId)) -> Option<(GraphId, NodeId)> {
-    Some(match kind {
-        EventKind::Realloc { new_ptr, .. } => return None,
-        EventKind::Alloc { ptr, .. } => return None,
-        EventKind::AddrOfLocal(ptr, _) => return None,
-        EventKind::Done => return None,
-        _ => obj,
-    })
-}
+impl EventKindExt for EventKind {
+    /// return the ptr of interest for a particular event
+    fn ptr(&self, _metadata: &EventMetadata) -> Option<Pointer> {
+        use EventKind::*;
+        Some(match *self {
+            CopyPtr(lhs) => lhs,
+            Field(ptr, ..) => ptr,
+            Free { ptr } => ptr,
+            Ret(ptr) => ptr,
+            LoadAddr(ptr) => ptr,
+            StoreAddr(ptr) => ptr,
+            LoadValue(ptr) => ptr,
+            StoreValue(ptr) => ptr,
+            CopyRef => return None, // FIXME
+            ToInt(ptr) => ptr,
+            Realloc { old_ptr, .. } => old_ptr,
+            FromInt(lhs) => lhs,
+            Alloc { ptr, .. } => ptr,
+            AddrOfLocal(lhs, _) => lhs,
+            Offset(ptr, _, _) => ptr,
+            Done => return None,
+        })
+    }
+
+    fn has_parent(&self) -> bool {
+        use EventKind::*;
+        match self {
+            Realloc { new_ptr: _, .. } => false,
+            Alloc { ptr: _, .. } => false,
+            AddrOfLocal(_ptr, _) => false,
+            Done => false,
+            _ => true,
+        }
+    }
 
-pub fn event_to_node_kind(event: &Event) -> Option<NodeKind> {
-    Some(match event.kind {
-        EventKind::Alloc { .. } => NodeKind::Malloc(1),
-        EventKind::Realloc { .. } => NodeKind::Malloc(1),
-        EventKind::Free { .. } => NodeKind::Free,
-        EventKind::CopyPtr(..) | EventKind::CopyRef => NodeKind::Copy,
-        EventKind::Field(_, field) => NodeKind::Field(field.into()),
-        EventKind::LoadAddr(..) => NodeKind::LoadAddr,
-        EventKind::StoreAddr(..) => NodeKind::StoreAddr,
-        EventKind::LoadValue(..) => NodeKind::LoadValue,
-        EventKind::StoreValue(..) => NodeKind::StoreValue,
-        EventKind::AddrOfLocal(_, l) => NodeKind::AddrOfLocal(Local::from(l)),
-        EventKind::ToInt(_) => NodeKind::PtrToInt,
-        EventKind::FromInt(_) => NodeKind::IntToPtr,
-        EventKind::Ret(_) => return None,
-        EventKind::Offset(_, offset, _) => NodeKind::Offset(offset),
-        EventKind::Done => return None,
-    })
+    fn parent(&self, obj: (GraphId, NodeId)) -> Option<(GraphId, NodeId)> {
+        if self.has_parent() {
```

Comment:
I think the logic here may be reversed. Why not have this defined the other way where `has_parent` is an alias for `parent(..).is_some()`?

Reply:
And by reversed I mean we should be returning `Some(obj)` if it has a parent

Reply:
Yeah, that is backwards, oops.

Reply:
I think it makes sense to have `parent` call `has_parent` because then `has_parent` doesn't need to take the `obj` argument.  If we had `has_parent` call `self.parent(obj).is_some()`, then we'd need that `obj`.

Reply:
Granted, I'm not sure I see the appeal of breaking them up this way though?

Reply:
I fixed the reversal now.

Reply:
I wasn't sure the appeal of having the `obj` as an argument, but I left that function there since there might've been other reasons for it.  But we could drop that and just do `.has_parent().then(|| obj)`.  To me, `has_parent` seems like it has the core logic.  `obj` doesn't really need to be in it at all.

Reply:
Yeah, I agree with you on the `obj` piece. Okay, let's have it this way, and we can find a better way to handle `obj` later on.